### PR TITLE
Implemented fix to Issue #258

### PR DESCRIPTION
**Pull request by huttered40**

### CODE REVIEW

**Comment by srajama1** [comment]
@huttered40 : Thanks for the changes ? Can you add a test as @kyungjoo-kim requested so we can merge this in.

**Comment by huttered40** [comment]
Yes, I'll add the test tomorrow.

**Comment by huttered40** [comment]
I added changes to the serial and team gemm unit tests, but I have not tested yet. Do you like the format Kyungjoo? When I hit 'make' within unit_test/ subdirectory, I get a bunch of compiler errors seemingly unrelated to my changes.

**Comment by huttered40** [comment]
I wanted to remove the changes made to the Makefile that I included in the previous commit by accident, but ended up removing the entire file with that latest commit. Any way to bring it back?

**Comment by srajama1** [comment]
@huttered40 : Once you confirm the unit tests pass, I will merge the code

**Comment by huttered40** [comment]
@srajama1 What is the best way to test? I am running 'make -j' within kokkos-kernels/unit_test/ and it hits a ton of type errors completely unrelated to my changes in the sparse gemm code and others. I am pretty sure there is nothing wrong with my Makefile, which I set to work with OpenMP and KNL since I am on Bowman.

**Comment by srajama1** [comment]
I run make test

**Comment by ndellingwood** [comment]
@huttered40 from your build directory try running `make install-lib` first, then move into the unit_test directory and try `make -j`, if you didn't try that step first. 
Also, could you post your lines for how you generated the makefile? If the steps above don't help I can take a look and try and help.

**Comment by huttered40** [comment]
@ndellingwood Here are my steps exactly:

1) Go to kokkos-kernels/example/buildlib/ directory
2) Modify compileKokkosKernels.sh to be the following:

KOKKOS_PATH=${HOME}/kokkoskernels_spgemm_benchmark/kokkos #path to kokkos source
KOKKOSKERNELS_SCALARS='double,"complex<double>"' #the scalar types to instantiate =double,float...
KOKKOSKERNELS_LAYOUTS=LayoutLeft #the layout types to instantiate.
KOKKOSKERNELS_ORDINALS=int,long #ordinal types to instantiate
KOKKOSKERNELS_OFFSETS=int,size_t #offset types to instantiate
KOKKOSKERNELS_PATH=../.. #path to kokkos-kernels top directory.
KOKKOSKERNELS_OPTIONS=eti-only #options for kokkoskernels  
CXXFLAGS="-Wall -pedantic -Werror -O3 -g -Wshadow -Wsign-compare -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized"
CXX=icpc #
KOKKOS_DEVICES=Serial,OpenMP #devices Cuda...
KOKKOS_ARCHS=KNL

../../scripts/generate_makefile.bash --kokkoskernels-path=${KOKKOSKERNELS_PATH} --with-scalars=${KOKKOSKERNELS_SCALARS} --with-ordinals=${KOKKOSKERNELS_ORDINALS} --with-offsets=${KOKKOSKERNELS_OFFSETS} --kokkos-path=${KOKKOS_PATH} --with-devices=${KOKKOS_DEVICES} --arch=${KOKKOS_ARCHS} --compiler=${CXX} --with-options=${KOKKOSKERNELS_OPTIONS}  --cxxflags="${CXXFLAGS}"

3) Run 'make install-lib -j'. I also separately tried 'make test-j' and same thing happened. I get a ton of compiler errors in files I didn't touch such as the following:

/ascldap/users/ehutter/KKissue258/kokkos-kernels/src/sparse/impl/KokkosSparse_spmv_impl.hpp(995): error: class "KokkosSparse::CrsMatrix<const Kokkos::complex<double>, const int, Kokkos::Device<Kokkos::Serial::execution_space, Kokkos::Serial::memory_space>, Kokkos::MemoryTraits<1U>, const size_t={unsigned long}>" has no member "const_ordinal_type"
      typename AMatrix::const_ordinal_type nrow = A.numRows();

I am running this on a KNL node with intel/compilers/18.0.128 module

Thanks for your help with this.

**Comment by ndellingwood** [comment]
@huttered40 instead of building in the example directory, try creating a subdirectory in kokkos-kernels and copying your `compileKokkosKernels.sh` script there and running. For example, I create a directory called `testing` and then a subdirectory for the type of build, e.g. `bowman-intel17-knl`, so the directory structure looks like this: `kokkos-kernels/testing/bowman-intel17-knl`
An important point you'll probably notice is that however you decide to create your subdirectory, make sure the level that you nest the subdirs is reflected in the path to the `generate_makefile.bash`, seen in the last line of the script `../../scripts/generate_makefile.bash`
Someone else had a similar issue and this helped resolve it for them, hopefully this is all it takes in your case as well. 
Also, so you don't get annoying untracked messages about your new subdir, you can add the outer-most directory to `.git/info/exclude` within the kokkos-kernels directory and it will be ignored.
Let me know if this gets you over this bump, if so we need to improve the documentation to prevent this issue.

**Comment by huttered40** [comment]
@ndellingwood I still get the same error. I do notice that after running 'make install-lib -j', a message pops up right before the avalanche of compiler errors saying "ls: cannot access KokkosKernels_config.h: No such file or directory". I assume that must be something bad, right? Hopefully that is what is causing the issue, although I don't know why it wouldn't be created, as I see it gets created in the Makefile.kokkos_kernels.

**Comment by huttered40** [comment]
@ndellingwood  Ok let me know if you can get it to build. It could just be something really dumb on my part, I'm not sure. The -j16 hint did not fix the errors.

**Comment by ndellingwood** [comment]
Hi @huttered40 

The minor mods on your script (pasted below) worked for me to get the library to build and so far so good building the unit tests using your PR. I am doing this on Bowman, with the same intel 18 module you listed. One difference is that my copy of Kokkos is located in the home directory, not that this should make any difference but mentioning so you reset it to your path. Adding quotes for some of the variables names in the script seems to have helped generate_makefile with parsing the input properly. 

```
KOKKOS_PATH=${HOME}/kokkos #path to kokkos source
KOKKOSKERNELS_SCALARS='double,"complex<double>"' #the scalar types to instantiate =double,float...
KOKKOSKERNELS_LAYOUTS=LayoutLeft #the layout types to instantiate.
KOKKOSKERNELS_ORDINALS="int,long" #ordinal types to instantiate
KOKKOSKERNELS_OFFSETS="int,size_t" #offset types to instantiate
KOKKOSKERNELS_PATH=../.. #path to kokkos-kernels top directory.
KOKKOSKERNELS_OPTIONS=eti-only #options for kokkoskernels
CXXFLAGS="-Wall -pedantic -Werror -O3 -g -Wshadow -Wsign-compare -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized"
CXX=icpc #
KOKKOS_DEVICES="Serial,OpenMP" #devices Cuda...
KOKKOS_ARCHS=KNL

../../scripts/generate_makefile.bash --kokkoskernels-path=${KOKKOSKERNELS_PATH} --with-scalars=${KOKKOSKERNELS_SCALARS} --with-ordinals=${KOKKOSKERNELS_ORDINALS} --with-offsets=${KOKKOSKERNELS_OFFSETS} --kokkos-path=${KOKKOS_PATH} --with-devices=${KOKKOS_DEVICES} --arch=${KOKKOS_ARCHS} --compiler=${CXX} --with-options=${KOKKOSKERNELS_OPTIONS} --cxxflags="${CXXFLAGS}"
```

**Comment by ndellingwood** [comment]
@huttered40 The tests built without error :)

I started running the OpenMP tests, and the following seg-faulted

```
[ RUN      ] openmp.graph_graph_color_d2_double_int_int_TestExecSpace
Segmentation fault (core dumped)
```

It doesn't look like your changes touched this test, though I'm not sure if it makes any calls to the gemm files you updated. I'm rerunning with this seg-faulted graph_color_d2 tests filtered out and will test a separate build of the develop branch to pin down if the seg-faulted test is pre-existing.

**Comment by ndellingwood** [comment]
For reference, the following tests failed (color d2 tests not run) but I will test on develop branch to make sure this is pre-existing:
```
[  FAILED  ] openmp.batched_scalar_serial_trsm_l_u_nt_n_dcomplex_dcomplex
[  FAILED  ] openmp.batched_scalar_serial_trsm_l_u_nt_n_dcomplex_double
[  FAILED  ] openmp.batched_scalar_team_trsm_l_u_nt_n_dcomplex_dcomplex
[  FAILED  ] openmp.batched_scalar_team_trsm_l_u_nt_n_dcomplex_double
```

**Comment by huttered40** [comment]
Ok good to know. Interestingly, I still have the same issues when using quotes in the Makefile as you suggested. I went back and tried building a different branch with the same Makefile and it built correctly, which would imply that something in my Issue258 branch is causing the issue, but you say that you got it to build correctly after pulling my branch. I'll keep looking at it.

**Comment by huttered40** [comment]
@ndellingwood Actually, to append my above comment, I tried building kokkos-kernels from scratch using the same approach as discussed above and I get the same exact compiler errors. I am using version 2.7.00 and the top commit was from Thursday May 24, "merge branch develop for 2.7.00". I built version 2.6.00 from scratch and it compiled fine however. So the master branch I was building my changes on I guess was never able to build in the first place, at least on the Bowman node I'm on.

**Comment by ndellingwood** [comment]
@huttered40 check your version of Kokkos is compatible with KokkosKernels - if you're using kokkos-kernels v 2.7.00 (or newer), kokkos should be v 2.7.00 (or version of the day (VOD) aka updated develop branch). If your version of kokkos is 2.6.00 (possibly with new merges but before 2.7.00) and kokkos-kernels is 2.7.00 there will be compilation issues.

**Comment by huttered40** [comment]
@ndellingwood Ah that must be it. I'll try again tomorrow morning.

**Comment by huttered40** [comment]
@ndellingwood  That fixed the build issue. Now when I run 'make test -j' I get an 'Illegal Instruction' in the OpenMP and Serial unit tests. Is that what you were seeing yesterday?

**Comment by srajama1** [comment]
@huttered40 You could see illegal instruction if you are not salloc'd.

The issue @ndellingwood are not due to your PR. See PR #249 where @lucbv saw these errors too. If everything else passes I can merge this.

**Comment by huttered40** [comment]
@srajama1 Ok, I'm in the middle of the tests now.

**Comment by huttered40** [comment]
@srajama1 @ndellingwood 

The failing tests I see are:

1) openmp.graph_graph_color_d2_double_int_int_TestExecSpace (segfault)
2) serial.batched_scalar_serial_trsm_l_u_nt_n_dcomplex_dcomplex (numerical error)
3) serial.batched_scalar_serial_trsm_l_u_nt_n_dcomplex_double (numerical error)
4) serial.batched_scalar_team_trsm_l_u_nt_n_dcomplex_dcomplex (numerical error)
5) serial.batched_scalar_team_trsm_l_u_nt_n_dcomplex_double (numerical error)

**Comment by kyungjoo-kim** [comment]
Merge with upstream/develop. If you didnot set up upstream, 
```
git remote add upstream ssh://git@github.com/kokkos/kokkos-kernels.git 
git fetch --all
git merge upstream/develop
```

**Comment by kyungjoo-kim** [comment]
@huttered40 You cannot commit the merging as it is. You need to resolve the conflicts.

**Comment by huttered40** [comment]
@kyungjoo-kim Not sure what you mean. I fixed the only conflict I had, as noted in my last commit message, and just re-build and re-tested everything and all 269 tests passed for both Serial and OpenMP. What other conflicts are you referring to?

File: src/batched/KokkosBatched_Gemm_Serial_Impl.hpp [hunk 1]
```diff
@@ -159,12 +159,13 @@ namespace KokkosBatched {
...
         m = C.dimension(0),
         n = C.dimension(1),
-        k = A.dimension(1);
+        k = A.dimension(0),
 
       static_assert(is_vector<vector_type>::value, "value type is not vector type");      
       static_assert(vector_type::vector_length == 4 || vector_type::vector_length == 8,
                     "AVX, AVX2 and AVX512 is supported");
       const MKL_COMPACT_PACK format = vector_type::vector_length == 8 ?  MKL_COMPACT_AVX512 : MKL_COMPACT_AVX;
+>>>>>>> upstream/develop
```
Comment (kyungjoo-kim) [inline src/batched/KokkosBatched_Gemm_Serial_Impl.hpp]: delete this stuffs.

Comment (huttered40) [reply src/batched/KokkosBatched_Gemm_Serial_Impl.hpp]: Ah missed htose last 2, my mistake.

Comment (huttered40) [reply src/batched/KokkosBatched_Gemm_Serial_Impl.hpp]: Hold on, I found some more issues.

Comment (huttered40) [reply src/batched/KokkosBatched_Gemm_Serial_Impl.hpp]: Obviously if the 'make install-lib' and 'make test' commands would have alerted me to these compiler errors, I would have fixed them before committing. I am not sure why it hasn't been detecting them and that I've had to eyeball them instead.

Comment (kyungjoo-kim) [reply src/batched/KokkosBatched_Gemm_Serial_Impl.hpp]: this code is instanciated when you enable mkl as its third party library. that is why you cannot catch during compiling.

Comment (huttered40) [reply src/batched/KokkosBatched_Gemm_Serial_Impl.hpp]: Ok. Strange, I didn't know that. That has been the root of all this trouble.

Comment (huttered40) [reply src/batched/KokkosBatched_Gemm_Serial_Impl.hpp]: Ah I see the MKL preprocessor ifdefs surrounding that method, makes sense to me.

**Comment by huttered40** [comment]
@srajama1 If @kyungjoo-kim agrees, I think this branch is ready to merge. The two main problems were that I was at first mistakenly using Kokkos-kernels version 2.7.00 with an old kokkos version 2.6.00, and once I got that fixed and initially merged, there were some lingering conflict statement brackets in a part of the code that was not being compiled (needed MKL), so that is why I didn't initially see it after building and testing.

**Comment by srajama1** [comment]
I am glad this was sorted out and thanks for the contribution to KK.

**Pull request merged**